zenpacklib.py

```diff
@@ -2661,6 +2661,26 @@ def datapoint_method(self, default=spec.datapoint_default, cached=spec.datapoint
                     }
                 catalogs[self.name]['indexes'].update(pindexes)
 
+        def global_only():
```

Comment:
Rather than adding all new code to fix the bad state catalogs has gotten into, I'd rather just replace the existing logic that's update catalogs to update it correctly to begin with. I think this would be a lot easier to follow when reading the code in the future, and this method is probably already too long to understand quickly.

Perhaps replacing the current _if pindexes:_ branch with this instead.

```python
# Add catalogs and indexes.
if pindexes:
    if self.name not in catalogs:
        catalogs[self.name] = {'indexes': {}}

    catalogs[self.name]['indexes'].update(pindexes)

    # Add a properly-scoped "id" index if it doesn't already exist.
    if 'id' not in catalogs[self.name]['indexes']:
        scopes = {
            x.get('scope', 'device')
            for x in catalogs[self.name]['indexes'].values()}

        catalogs[self.name]['indexes']['id'] = {
            'type': 'field',
            'scope': "device" if "device" in scopes else "global"}
```